columnflow/plotting/plot_all.py

```diff
@@ -8,51 +8,154 @@
+        "edgecolor": "black",
         "alpha": 1.0,
     }
-    defaults.update(kwargs)
-    ax.bar(**defaults)
+    ax.bar(**bar_kwargs)
+
+
+def draw_syst_error_bands(
+    ax: plt.Axes,
+    h: hist.Hist,
+    syst_hists: Sequence[hist.Hist],
+    shift_insts: Sequence[od.Shift],
+    norm: float | Sequence | np.ndarray = 1.0,
+    method: str = "quadratic_sum",
+    **kwargs,
+) -> None:
+    assert len(h.axes) == 1
+    assert method in ("quadratic_sum", "envelope")
+
+    nominal_shift, shift_groups = group_shifts(shift_insts)
+    if nominal_shift is None:
+        raise ValueError("no nominal shift found in the list of shift instances")
+
+    # create pairs of shifts mapping from up -> down and vice versa
+    shift_pairs = {}
+    for up_shift, down_shift in shift_groups.values():
+        shift_pairs[up_shift] = down_shift
+        shift_pairs[down_shift] = up_shift
+
+    # stack histograms separately per shift, falling back to the nominal one when missing
+    shift_stacks: dict[od.Shift, hist.Hist] = {}
+    for shift_inst in sum(shift_groups.values(), []):
+        for _h in syst_hists:
+            # when the shift is present, the flipped shift must exist as well
+            shift_ax = _h.axes["shift"]
+            if shift_inst.name in shift_ax:
+                if shift_pairs[shift_inst].name not in shift_ax:
+                    raise RuntimeError(
+                        f"shift {shift_inst} found in histogram but {shift_pairs[shift_inst]} is missing; "
+                        f"existing shifts: {','.join(map(str, list(shift_ax)))}",
+                    )
+                shift_name = shift_inst.name
+            else:
+                shift_name = nominal_shift.name
+            # store the slice
+            _h = _h[{"shift": hist.loc(shift_name)}]
+            if shift_inst not in shift_stacks:
+                shift_stacks[shift_inst] = _h
+            else:
+                shift_stacks[shift_inst] += _h
+
+    # loop over bins, subtract nominal yields from stacked yields and merge differences into
+    # a systematic error per bin using the given method (quadratic sum vs. evelope)
+    # note 1: if the up/down variations of the same shift source point in the same direction, a
+    #         statistical combination is pointless and their minimum/maximum is selected instead
+    # note 2: relative signs are consumed into the meaning of "up" and "down" here as they already
+    #         are combinations evaluated for a specific direction
+    syst_error_up = []
+    syst_error_down = []
+    for b in range(h.axes[0].size):
+        up_diffs = []
+        down_diffs = []
+        for source, (up_shift, down_shift) in shift_groups.items():
+            up_diff = shift_stacks[up_shift].values()[b] - h.values()[b]
+            down_diff = shift_stacks[down_shift].values()[b] - h.values()[b]
+            up_diffs.append(max(up_diff, down_diff, 0))
+            down_diffs.append(min(up_diff, down_diff, 0))
```

Comment:
maybe rename variable names for more verbosity? e.g.
`up_diff` --> `shift_up_diff`
`up_diffs` --> `value_up_diffs`

Reply:
I'm not sure if I'd understand the difference between `shift_` and `value_` here.

Reply:
I agree with you, I just didn't come up with anything better yet.
I'd appreciate to somehow separate the "shift diffs" (`up` can both increase and decrease) and "value diffs" (`up` always means that the value increases).

Reply:
Yes, thinking about it, I agree.

In 580d0d79 I added comments and renamed `{up,down}_diff` to `shift_{up,down}_diff`.